23.0.0.7 staging (#3230)
OpenLiberty/blogs: posts/2023-07-25-23.0.0.7.adoc

at OpenTracingPropagatorsBuilder.<init>(OpenTracingPropagatorsBuilder.java:16)
	at OpenTracingPropagators.builder(OpenTracingPropagators.java:32)
	at OpenTracingShim.createTracerShim(OpenTracingShim.java:61)
	at OpenTracingShimServlet.createShim(OpenTracingShimServlet.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
----
+
This issue now resolved and calling a method on `GlobalOpenTelemetry` delegates to the no-op `OpenTelemetry` class.

> * link:https://github.com/OpenLiberty/open-liberty/issues/25479[Unable to make field long java.nio.Buffer.address accessible when using Java 17]

Reviewer: ```suggestion
```

---
layout: post
title: "API and SPI Javadoc in Open Liberty 23.0.0.7"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/mbroz2
author_github: https://github.com/mbroz2
seo-title: API and SPI Javadoc in Open Liberty 23.0.0.7 - OpenLiberty.io

> seo-description: Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs available via the Open Liberty docs page have been expended to include the Liberty specific APIs and SPIs.  This will make it easier to develop applications and custom features that make use of various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs have been fixed as part of this release.

Reviewer 2: ```suggestion
seo-description: Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs in the Open Liberty Docs are expanded to include the Liberty-specific APIs and SPIs.  This update simplifies developing applications and custom features that use various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs were fixed as part of this release.
```

> seo-description: Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs available via the Open Liberty docs page have been expended to include the Liberty specific APIs and SPIs.  This will make it easier to develop applications and custom features that make use of various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs have been fixed as part of this release.

Reviewer 2: ```suggestion
seo-description: Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs in the Open Liberty Docs are expanded to include the Liberty-specific APIs and SPIs.  This update simplifies developing applications and custom features that use various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs were fixed as part of this release.
```

Reviewer: duplicate of the first suggestion. Marking this one as resolved

> blog_description: Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs available via the Open Liberty docs page have been expended to include the Liberty specific APIs and SPIs.  This will make it easier to develop applications and custom features that make use of various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs have been fixed as part of this release.

Reviewer 2: ```suggestion
blog_description: Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs in the Open Liberty Docs are expanded to include the Liberty-specific APIs and SPIs.  This update simplifies developing applications and custom features that use various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs were fixed as part of this release.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= API and SPI Javadoc in Open Liberty 23.0.0.7
Michal Broz <https://github.com/mbroz2>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs available via the Open Liberty docs page have been expended to include the Liberty specific APIs and SPIs.  This will make it easier to develop applications and custom features that make use of various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs have been fixed as part of this release.

Reviewer 2: ```suggestion
Concurrent with the Open Liberty 23.0.0.7 release, the Javadocs in the Open Liberty docs are expanded to include the Liberty-specific APIs and SPIs.  This update simplifies developing applications and custom features that use various Liberty capabilities, like OpenID Connect and OAuth. Also, various bugs were fixed as part of this release.
```

In link:{url-about}[Open Liberty] 23.0.0.7:

* <<javadoc, Liberty API and SPI Javadoc>>
* <<bugs, Notable bug fixes>>

Check out link:{url-prefix}/blog/?search=release&search!=beta[previous Open Liberty GA release blog posts].

[#run]
== Run your apps using 23.0.0.7

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>23.0.0.7</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[23.0.0.7,)'
}
----

Or if you're using link:{url-prefix}/docs/latest/container-images.html[container images]:

[source]
----
FROM icr.io/appcafe/open-liberty
----

Or take a look at our link:{url-prefix}/start/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#javadoc]
== Liberty API and SPI Javadoc

> Prior to the Open Liberty 23.0.0.7 release, when working with Liberty APIs and SPIs developers had to use the embedded Javadoc that comes packaged with the runtime (or build their own from source).  Starting with 23.0.0.7, we've also made the Javadoc available on the Open Liberty documentation page, alongside the other Javadoc for MicroProfile, Jakarta EE and Java EE.  We've also retroactively made the Javadoc available for 23.0.0.6.

Reviewer 2: ```suggestion
Before the Open Liberty 23.0.0.7 release, developers who were working with Liberty APIs and SPIs had to use the embedded Javadoc that comes packaged with the runtime or build their own Javadocs from source.  Starting with 23.0.0.7, we've added the API and SPI Javadoc to the Open Liberty Docs, alongside the Javadoc for MicroProfile, Jakarta EE, and Java EE.  We've also retroactively added the Javadoc for 23.0.0.6.
```

> To start using the hosted Javadoc, navigate to the REFERENCE section of the link:https://openliberty.io/docs/latest/overview.html[docs]. 

Reviewer 2: ```suggestion
To start using the hosted Javadoc, navigate to the **REFERENCE** section of the link:https://openliberty.io/docs/latest/overview.html[Open Liberty Docs]. 
```

at OpenTracingPropagatorsBuilder.<init>(OpenTracingPropagatorsBuilder.java:16)
	at OpenTracingPropagators.builder(OpenTracingPropagators.java:32)
	at OpenTracingShim.createTracerShim(OpenTracingShim.java:61)
	at OpenTracingShimServlet.createShim(OpenTracingShimServlet.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
----
+
This issue now resolved and calling a method on `GlobalOpenTelemetry` delegates to the no-op `OpenTelemetry` class.

* link:https://github.com/OpenLiberty/open-liberty/issues/25479[Unable to make field long java.nio.Buffer.address accessible when using Java 17]
+

> During server start with `com.ibm.ws.tcpchannel.useZosAio=true` specified, an FFDC is created when running with Java 17 on z/OS:

Reviewer 2: do we post z/Os bugs on Open Liberty? My understanding was OL technically doesn't support running on z/Os

> During server start with `com.ibm.ws.tcpchannel.useZosAio=true` specified, an FFDC is created when running with Java 17 on z/OS:
> +
> [source]
> ----
> Exception = java.lang.RuntimeException
> Source = com.ibm.ws.tcpchannel.internal.NBAcceptChannelSelector
> probeid = 254
> Stack Dump = java.lang.RuntimeException: java.lang.NoClassDefFoundError: com.ibm.io.async.AbstractAsyncChannel (initialization failure)
> Caused by: java.lang.NoClassDefFoundError: com.ibm.io.async.AbstractAsyncChannel (initialization failure)
> Caused by: java.lang.RuntimeException: java.lang.reflect.InaccessibleObjectException: Unable to make field long java.nio.Buffer.address accessible: module java.base does not "opens java.nio" to unnamed module @299d0dca
> Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field long java.nio.Buffer.address accessible: module java.base does not "opens java.nio" to unnamed module @299d0dca
> ----
> +
> This issue has been resolved and the FFDC is no longer produced.
> 

Reviewer: ```suggestion
```

* link:https://github.com/OpenLiberty/open-liberty/issues/19861[Concurrency errors when using same JWT access token for inbound propagation]
+

> For a Liberty server configured with `openidConnectClient-1.0` using the same JWT access token for inbound propagation for multiple requests, there is a concurrency error which causes the requests to fail.  This results in a `ConcurrentModificationException` occurring:

Reviewer 2: ```suggestion
For a Liberty server configured with the`openidConnectClient-1.0` feature and using the same JWT access token for inbound propagation for multiple requests, a concurrency error occurs that causes the requests to fail.  This error results in a `ConcurrentModificationException`:
```

+
[source]
----
Stack Dump = java.util.ConcurrentModificationException
	at java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:977)
	at java.util.LinkedList$ListItr.next(LinkedList.java:899)
	at javax.security.auth.Subject$SecureSet$1.next(Subject.java:1290)
	at com.ibm.ws.security.openidconnect.client.internal.AccessTokenCacheHelper.recreateSubject(AccessTokenCacheHelper.java:137)
	at com.ibm.ws.security.openidconnect.client.internal.AccessTokenCacheHelper.getCachedTokenAuthenticationResult(AccessTokenCacheHelper.java:51)
	at com.ibm.ws.security.openidconnect.client.AccessTokenAuthenticator.authenticate(AccessTokenAuthenticator.java:130)
	at com.ibm.ws.security.openidconnect.client.internal.OidcClientImpl.authenticate(OidcClientImpl.java:406)
----
+

> This issue has been resolved; the `ConcurrentModificationException` is not longer thrown and the requests succeed.

Reviewer 2: ```suggestion
This issue has been resolved. The `ConcurrentModificationException` is not longer thrown and the requests succeed.
```
